diagnostics/reporter.rs

```diff
@@ -0,0 +1,154 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    fs,
+    hash::{DefaultHasher, Hash, Hasher},
+    io::{self, Write},
+    path::{Path, PathBuf},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc, Mutex,
+    },
+    thread,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use chrono::{DateTime, Local, Timelike};
+use concurrency::IntervalRunner;
+use logger::{debug, trace};
+use reqwest::{
+    blocking::{Client, Response},
+    header::{HeaderValue, CONNECTION, CONTENT_TYPE},
+};
+use resource::constants::diagnostics::{
+    DATABASE_METRICS_UPDATE_INTERVAL, DISABLED_REPORTING_FILE_NAME, REPORT_INTERVAL,
+};
+
+use crate::Diagnostics;
+
+#[derive(Debug)]
+pub struct Reporter {
+    deployment_id: String,
+    diagnostics: Arc<Diagnostics>,
+    reporting_uri: &'static str,
+    data_directory: PathBuf,
+    is_enabled: bool,
+    _reporting_job: Arc<Option<IntervalRunner>>,
+}
+
+impl Reporter {
+    pub(crate) fn new(
+        deployment_id: String,
+        diagnostics: Arc<Diagnostics>,
+        reporting_uri: &'static str,
+        data_directory: PathBuf,
+        is_enabled: bool,
+    ) -> Self {
+        Self { deployment_id, diagnostics, reporting_uri, data_directory, is_enabled, _reporting_job: Arc::new(None) }
+    }
+
+    pub fn may_start(&self) {
+        if self.is_enabled {
+            Self::delete_disabled_reporting_file_if_exists(&self.data_directory);
+            self.schedule_reporting();
+        } else {
+            self.report_once_if_needed();
+        }
+    }
+
+    fn schedule_reporting(&self) {
+        let diagnostics = self.diagnostics.clone();
+        let reporting_uri = self.reporting_uri;
+
+        let reporting_job = IntervalRunner::new_with_initial_delay(
```

Comment:
Right now, I use an IntervalRunner, which uses an `std::thread::spawn` under the hood.